File: Lab 3/Java/ProblemA1.java
import java.util.*;

public class ProblemA1 {
    static long inversions = 0;

    static int[] merge(int[] a, int[] b) {
        int[] c = new int[a.length + b.length];
        int i = 0, j = 0, k = 0;
        long inv = 0;
        while (i < a.length && j < b.length) {
            if (a[i] > b[j]) {
                c[k++] = b[j++];
                inv += (a.length - i);
            } else {
                c[k++] = a[i++];
            }
        }
        while (i < a.length)
            c[k++] = a[i++];
        while (j < b.length)
            c[k++] = b[j++];
        inversions += inv;
        return c;
    }

    static int[] mergeSort(int[] arr) {
        if (arr.length <= 1)
            return arr;
        int mid = arr.length / 2;
        int[] left = Arrays.copyOfRange(arr, 0, mid);
        int[] right = Arrays.copyOfRange(arr, mid, arr.length);
        int[] sortedLeft = mergeSort(left);
        int[] sortedRight = mergeSort(right);
        System.out.println(Arrays.toString(sortedLeft) + " " + Arrays.toString(sortedRight));
        int[] result = merge(sortedLeft, sortedRight);
        System.out.println(Arrays.toString(result));
        return result;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (!sc.hasNextInt())
            return;
        int n = sc.nextInt();
        int[] arr = new int[n];
        for (int i = 0; i < n; i++)
            arr[i] = sc.nextInt();
        int[] res = mergeSort(arr);
        System.out.println(inversions);
        for (int val : res)
            System.out.print(val + " ");
        System.out.println();
    }
}

File: Lab 3/ProblemA1.py
import sys
input = sys.stdin.readline
def merge(a,b):
    inversions = 0
    i,j = 0,0
    c = []
    while i < len(a) and j < len(b):
        if a[i] > b[j]:
            c.append(b[j])            
            j += 1
            inversions += len(a) - i
        else:
            c.append(a[i])
            i += 1
    c.extend(a[i:])
    c.extend(b[j:])
    return c, inversions

def mergesort(arr):
    if len(arr) <= 1:
        return arr, 0
    mid = len(arr)//2
    a1, left = mergesort(arr[:mid])
    a2, right = mergesort(arr[mid:])
    print(a1, a2)
    result, inversion = merge(a1,a2)
    print(result)
    return result, left + right + inversion

    
length = int(input())

array = list(map(int, input().split()))

result, inversions = mergesort(array)

print(inversions)
print(*result)